subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaBasePlugin.java

```diff
@@ -378,37 +377,45 @@ public Object call() throws Exception {
-        buildTask.setDescription("Assembles and tests this project and all projects it depends on.");
-        buildTask.setGroup(BasePlugin.BUILD_GROUP);
-        buildTask.dependsOn(BUILD_TASK_NAME);
+        project.getTasks().createLater(BUILD_NEEDED_TASK_NAME, new Action<Task>() {
+            @Override
+            public void execute(Task buildTask) {
+                buildTask.setDescription("Assembles and tests this project and all projects it depends on.");
+                buildTask.setGroup(BasePlugin.BUILD_GROUP);
+                buildTask.dependsOn(BUILD_TASK_NAME);
+            }
+        });
     }
 
     private void configureBuildDependents(Project project) {
-        DefaultTask buildTask = project.getTasks().create(BUILD_DEPENDENTS_TASK_NAME, DefaultTask.class);
-        buildTask.setDescription("Assembles and tests this project and all projects that depend on it.");
-        buildTask.setGroup(BasePlugin.BUILD_GROUP);
-        buildTask.dependsOn(BUILD_TASK_NAME);
-        buildTask.doFirst(new Action<Task>() {
+        project.getTasks().createLater(BUILD_DEPENDENTS_TASK_NAME, new Action<Task>() {
             @Override
-            public void execute(Task task) {
-                if (!task.getProject().getGradle().getIncludedBuilds().isEmpty()) {
-                    task.getProject().getLogger().warn("[composite-build] Warning: `" + task.getPath() + "` task does not build included builds.");
-                }
+            public void execute(Task buildTask) {
+                buildTask.setDescription("Assembles and tests this project and all projects that depend on it.");
+                buildTask.setGroup(BasePlugin.BUILD_GROUP);
+                buildTask.dependsOn(BUILD_TASK_NAME);
+                buildTask.doFirst(new Action<Task>() {
+                    @Override
+                    public void execute(Task task) {
+                        if (!task.getProject().getGradle().getIncludedBuilds().isEmpty()) {
+                            task.getProject().getLogger().warn("[composite-build] Warning: `" + task.getPath() + "` task does not build included builds.");
+                        }
+                    }
+                });
             }
         });
     }
 
     private void configureTest(final Project project, final JavaPluginConvention convention) {
-        project.getTasks().withType(Test.class, new Action<Test>() {
+        project.getTasks().configureEachLater(Test.class, new Action<Test>() {
             public void execute(final Test test) {
                 configureTestDefaults(test, project, convention);
             }
         });
         project.getGradle().getTaskGraph().whenReady(new Action<TaskExecutionGraph>() {
             @Override
             public void execute(final TaskExecutionGraph taskExecutionGraph) {
-                project.getTasks().withType(Test.class, new Action<Test>() {
+                project.getTasks().configureEachLater(Test.class, new Action<Test>() {
```

Comment:
Probably not necessary to be lazy here since this is after the task graph is constructed, so the final set of tasks should already be set.